controllers/catalog_predicates.go

```diff
@@ -0,0 +1,115 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	catalogd "github.com/operator-framework/catalogd/pkg/apis/core/v1beta1"
+	operatorv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Predicate for reconciling operators when available (i.e. ready) catalogsources on cluster change
+type catalogReadyTransitionPredicate struct {
+	predicate.Funcs
+	catalogReady map[string]bool
+}
+
+func newCatalogReadyTransitionPredicate() *catalogReadyTransitionPredicate {
+	return &catalogReadyTransitionPredicate{
+		catalogReady: map[string]bool{},
+	}
+}
+
+func (c *catalogReadyTransitionPredicate) Create(e event.CreateEvent) bool {
+	fmt.Println("CreateEvent CatalogSource", e.Object.GetName())
+	catalogReady, err := isCatalogReady(e.Object)
+	if err != nil {
+		fmt.Println(err)
+		return false
+	}
```

Comment:
If the catalog is not ready, is that something we would want to reflect in the status of the Operator (either now, or in the future)?

I _think_ I'm of the opinion that we should just blindly forward along all CatalogSource events and deal with the state of it during reconciliation where we have a change to touch the Operator status if necessary.

WDYT?

Reply:
Depending on whether we have a caching layer for catalog contents at the operator controller level, we may not want to reflect a non-ready state for catalogs at the operator-controller level. I don't think it makes sense to run the resolver to trigger on any catalog change especially if the catalog is not ready, but I agree that we can have that check moved further down the resolution process. We'd need to verify that none of our sources are stale/non-ready to avoid resolving unnecessarily.